analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs

```diff
@@ -257,13 +271,23 @@ private static NumberConstraint AccountForZero(NumberConstraint constraint)
 
     private static BigInteger? CalculateXorMax(NumberConstraint left, NumberConstraint right)
     {
-        if ((left.IsPositive && right.CanBePositive) || (right.IsPositive && left.CanBeNegative))
+        if ((left.IsPositive && right.IsNegative) || (left.IsNegative && right.IsPositive))
         {
-            return left.Max.HasValue && right.Max.HasValue ? PositiveMagnitude(left.Max.Value | right.Max.Value) : null;
+            return -1;
         }
-        else if ((left.IsPositive && right.IsNegative) || (left.IsNegative && right.IsPositive))
+        else if ((left.IsPositive || right.IsPositive) && left.Max.HasValue && right.Max.HasValue)
         {
-            return -1;
+            return PositiveMagnitude(BigInteger.Max(left.Max.Value, right.Max.Value));
+        }
+        else if ((left.IsNegative || right.IsNegative) && left.Min.HasValue && right.Min.HasValue)
+        {
+            return PositiveMagnitude(BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value)));
+        }
+        else if (left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue)
+        {
+            return PositiveMagnitude(BigInteger.Max(
```

Comment:
Does the multi-level max have semantics here, or it's just lack of a Max of more than 2 values?
If it has semantics: a comment would help understanding what that is about.
If it hasn't, what about a private static method `Max` with params args, or the max of an array?